sympy/functions/combinatorial/numbers.py

```diff
@@ -1453,6 +1456,116 @@ def _eval_evalf(self, prec):
+
+    .. math :: \operatorname{E}_{2k} = (-1)^k \mathcal{A}_{2k}
+
+    Like the Bernoulli and Euler numbers, the Andre numbers are interpolated by the
+    entire Andre function:
+
+    .. math :: \mathcal{A}(s) = (-i)^{s+1} \operatorname{Li}_{-s}(i) +
+            i^{s+1} \operatorname{Li}_{-s}(-i) = \\ \frac{2 \Gamma(s+1)}{(2\pi)^{s+1}}
+            (\zeta(s+1, 1/4) - \zeta(s+1, 3/4) \cos{\pi s})
+
+    Examples
+    ========
+
+    >>> from sympy import andre, euler, bernoulli
+    >>> [andre(n) for n in range(11)]
+    [1, 1, 1, 2, 5, 16, 61, 272, 1385, 7936, 50521]
+    >>> [(-1)**k * andre(2*k) for k in range(7)]
+    [1, -1, 5, -61, 1385, -50521, 2702765]
+    >>> [euler(2*k) for k in range(7)]
+    [1, -1, 5, -61, 1385, -50521, 2702765]
+    >>> [andre(2*k-1) * (2*k) / ((-4)**k - (-16)**k) for k in range(1, 8)]
+    [1/6, -1/30, 1/42, -1/30, 5/66, -691/2730, 7/6]
+    >>> [bernoulli(2*k) for k in range(1, 8)]
+    [1/6, -1/30, 1/42, -1/30, 5/66, -691/2730, 7/6]
+
+    See Also
+    ========
+
+    bernoulli, catalan, euler, sympy.polys.appellseqs.andre_poly
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Alternating_permutation
+    .. [2] https://mathworld.wolfram.com/EulerZigzagNumber.html
+    .. [3] Peter Luschny, "An introduction to the Bernoulli function",
+           https://arxiv.org/abs/2009.06743
+    """
+
+    @classmethod
+    def eval(cls, n):
+        if n is S.NaN:
+            return S.NaN
+        elif n is S.Infinity:
+            return S.Infinity
+        if n.is_zero:
+            return S.One
+        elif n == -1:
+            return -log(2)
+        elif n == -2:
+            return -2*S.Catalan
+        elif n.is_Integer:
+            if n.is_nonnegative and n.is_even:
+                return abs(euler(n))
+            elif n.is_odd:
+                from sympy.functions.special.zeta_functions import zeta
+                m = -n-1
+                return I**m * Rational(1-2**m, 4**m) * zeta(-n)
+
+    def _eval_rewrite_as_zeta(self, s, **kwargs):
+        from sympy.functions.elementary.trigonometric import cos
+        from sympy.functions.special.gamma_functions import gamma
+        from sympy.functions.special.zeta_functions import zeta
+        return 2 * gamma(s+1) / (2*pi)**(s+1) * \
+                (zeta(s+1, S.One/4) - cos(pi*s) * zeta(s+1, S(3)/4))
+
+    def _eval_rewrite_as_polylog(self, s, **kwargs):
+        from sympy.functions.special.zeta_functions import polylog
+        return (-I)**(s+1) * polylog(-s, I) + I**(s+1) * polylog(-s, -I)
+
+    def _eval_is_integer(self):
+        n = self.args[0]
+        if n.is_integer and n.is_nonnegative:
+            return True
+
+    def _eval_is_positive(self):
+        if self.args[0].is_nonnegative:
+            return True
+
+    def _eval_evalf(self, prec):
+        from sympy.functions.special.zeta_functions import polylog
+        return self.rewrite(polylog)._eval_evalf(prec)
```

Comment:
This has come up in several PRs and I haven't raised this before but I've been trying to think what's a better way of handling evalf for these functions. I think this is the point that @asmeurer raised somewhere but really these should be using `_eval_mpmath` rather than `_eval_evalf`.

There are several problems with using `rewrite` here:

1. It's slow compared to a direct mpmath-based function.
2. The `rewrite` should at least use `deep=False` to avoid recursive rewrites for subexpressions.
3. The `rewrite` invokes symbolic as well as numeric evaluation so a bug in one leads to a bug in both.

Ideally there would just be a straight-forward mpmath function for each of these that could be used directly. Failing that at least somewhere in the sympy codebase I would want to see a function that shows how to compute this using mpmath (with the hope of one day upstreaming some form of it to mpmath).

I think the rewrite method makes sense for lambdify because there the rewrite is done once but used many times. Here every individual call to evalf triggers a symbolic rewrite which seems like too much.

Reply:
My take on this is that `_eval_mpmath` should be used if there's a single mpmath function for the given function (or maybe a different one depending on the number of arguments, like with `euler`). Actually if it's just a single function translation it's better to put it in the `MPMATH_TRANSLATIONS` dictionary in `lambdify.py`, because that will also make it work with `lambdify` (which is needed, e.g., for `nsolve`). 

If the function has to be evaluated as a larger expression, `_eval_mpmath` is not appropriate, because, as I showed in https://github.com/sympy/sympy/issues/23995#issuecomment-1232320702, it can bypass some of the SymPy code that handles automatic increasing of precision. In those cases, I'd personally rather see it done the way it was here before rather than the way it is being done now. IMO very few custom `_eval_evalf` methods should be calling mpmath directly. Rather, they should be able to reuse the logic that's already in the core `_eval_evalf` methods and `evalf.py`. 

If rewriting in these evalf methods is slow, that's something that we should consider if we can fix globally in the design on evalf.

Reply:
> If the function has to be evaluated as a larger expression, `_eval_mpmath` is not appropriate

Maybe this needs more thought.